File: credit.c
#include <cs50.h>
#include <stdio.h>

int main(void)
{
    long number = get_long("Number: ");
    long number2 = number;
    long number3 = number;


    int sum = 0; // sum of digits of products of 2 and every other digit of number
    int thisNum;

    number = number/10; // start with 2nd last digit
    while(number>0)
    {
        thisNum = number%10;
        thisNum = thisNum * 2;
        while(thisNum>0)
        {
            sum += (thisNum%10);
            thisNum = thisNum/10;
        }

        number = number/100;
    }

    while(number2>0)
    {
        sum += number2 % 10;

        number2 = number2 / 100;
    }

    bool valid;

    if(sum%10 == 0)
    {
        valid = true;
    }
    else
    {
        valid = false;
    }



    if(!valid)
    {
        printf("INVALID\n");
    }
    else
    {
        while(number3>99)
        {
            number3 = number3 / 10;
        }

        if(number3 == 34 || number3 == 37)
        {
            printf("AMEX\n");
        }
        else if(number3 >= 51 && number3 <=55)
        {
            printf("MASTERCARD\n");
        }
       else  if(number3 / 10 == 4)
        {
            printf("VISA\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }

}

File: credit.py
from cs50 import get_string

def main():
    number = get_string("Number: ")
    intnumber = int(number)
    isCard = valid(intnumber)
    if isCard == False:
        print("INVALID")
        exit(1)
    findtype(intnumber)
    exit(0)


def valid(num):
    num2 = num
    sum = 0
    # start with 2nd-to-last digit
    num //= 10
    started = False
    while num > 0:
        # get the end digit times 2
        thisNum = num%10
        thisNum *= 2
        # get the sum of the digits of (end digit * 2)
        while thisNum > 0:
            sum += thisNum%10
            thisNum //= 10
        # every other digit
        num //= 100
    # add the sum of the digits that weren't multiplied by 2
    while num2 > 0:
        sum += num2%10
        num2 //= 100
    # valid if the end sum is divisible by 10
    if sum % 10 == 0:
        return True
    return False

def findtype(num):
    digits = 0
    # find first 2 digits (when reading from left to right)
    while num > 99:
        num //= 10
        digits += 1
    if (num == 34 or num == 37) and digits+2 == 15:
        print("AMEX")
    elif (num >= 51 and num <= 55) and digits+2 == 16:
        print("MASTERCARD")
    elif num // 10 == 4 and (digits+2 == 13 or digits+2 == 16):
        print("VISA")
    else:
        print("INVALID")

main()
